dali/core/mm/default_resources.cc

```diff
@@ -0,0 +1,282 @@
+      abandon(p);
+    } else {
+      p = {};
+    }
+  }
+
+  template <typename T>
+  void abandon(std::vector<T> &v) {
+    for (auto &x : v)
+      abandon(x);
+    v.clear();
+  }
+
+  template <typename T>
+  void abandon(std::shared_ptr<T> &p) {
+    // shared_ptr doesn't have a release() function like one in unique_ptr so we could
+    // safely abandon the pointer.
+    // What happens here: we create (by placement new) a shared pointer and move the original
+    // pointer there - this way the original pointer is released and the new pointer
+    // is allocated dynamically on stack and never destroyed.
+    // This accomplishes the intentional leaking of the pointer.
+    using Ptr = std::shared_ptr<T>;
+    std::aligned_storage_t<sizeof(Ptr), alignof(Ptr)> dump;
+    new (&dump) Ptr(std::move(p));
+  }
+};
+
+#define g_resources (DefaultResources::instance())
+
+
+inline std::shared_ptr<host_memory_resource> CreateDefaultHostResource() {
+  return std::make_shared<malloc_memory_resource>();
+}
+
+struct CUDARTLoader {
+  CUDARTLoader() {
+    int device_id = 0;
+    CUDA_CALL(cudaGetDevice(&device_id));
+    DeviceGuard dg(device_id);
+  }
+};
+
+
+template <class Callable>
+struct CallAtExit {
+  explicit CallAtExit(Callable &&c) : callable(std::move(c)) {}
+  ~CallAtExit() {
+    callable();
+  }
+  Callable callable;
+};
+
+template <typename Callable>
+CallAtExit<Callable> AtExit(Callable &&c) {
+  return CallAtExit<Callable>(std::forward<Callable>(c));
+}
+
+inline std::shared_ptr<device_async_resource> CreateDefaultDeviceResource() {
+  static CUDARTLoader CUDAInit;
+#ifdef DALI_USE_RMM_DEVICE_RESOURCE
+  auto upstream = std::make_shared<rmm::mr::cuda_memory_resource>();
+  return rmm::mr::make_owning_wrapper<rmm::mr::pool_memory_resource>(std::move(upstream));
+#else
+  using resource_type = mm::async_pool_resource<mm::memory_kind::device>;
+  static auto upstream = std::make_shared<mm::cuda_malloc_memory_resource>();
+  auto rsrc = std::make_shared<resource_type>(upstream.get());
+  return make_shared_composite_resource(std::move(rsrc), std::move(upstream));
+#endif
+}
+
+inline std::shared_ptr<pinned_async_resource> CreateDefaultPinnedResource() {
+  using resource_type = mm::async_pool_resource<mm::memory_kind::pinned>;
+  auto upstream = std::make_shared<rmm::mr::pinned_memory_resource>();
+  auto rsrc = std::make_shared<resource_type>(upstream.get());
+  return make_shared_composite_resource(std::move(rsrc), std::move(upstream));
+}
+
+template <memory_kind kind>
+const std::shared_ptr<default_memory_resource_t<kind>> &ShareDefaultResourceImpl();
+
+template <>
+const std::shared_ptr<host_memory_resource> &ShareDefaultResourceImpl<memory_kind::host>() {
+  if (!g_resources.host)
+    g_resources.host = CreateDefaultHostResource();
+  return g_resources.host;
+}
+
+template <>
+const std::shared_ptr<pinned_async_resource> &ShareDefaultResourceImpl<memory_kind::pinned>() {
+  if (!g_resources.pinned_async) {
```

Comment:
Is it valid to Get the resource (so the default one is created) and then switch it to something else.
The AtExit is only used if the resource is not created. Does it miss any cases?

Reply:
It's valid. If on first access we create the resources, they can still be replaced with something else.
The AtExit makes sure that we release the resources sooner than the whole placeholder is destroyed, which in turn guarantees that CUDA runtime is not unloaded before.
The following order caused problems:
`GetDefaultResource<host>`  - initializes the g_resources
`GetDefaultResource<device>` (or pinned) - initializes CUDA runtime - therefore, normally the runtime would unload before the destruction of g_resources - this caused hard errors in RMM, because the pool resource tried to release the memory pool upon destruction, but the upstream resource called CUDA free and caused an error.